RIPS/rip-9995.md

```diff
@@ -0,0 +1,199 @@
+```
+
+## Rationale
+
+### ABI encoding for system frames' input and output data
+
+Using an encoding format for data provided by the protocol itself does represent an abstraction levels violation,
+it is a relatively safe one any alternative solution would require some trade-offs.
+
+The validation section of a Smart Contract Account code needs to have full access to the majority of transaction
+details in order to be able to make an informed decision about either accepting or rejecting the transaction.
+
+A small subset of this data is available with the existing opcodes, however creating an opcode for every transaction
+parameter is not feasible.
+
+Allowing wallets to specify their own encoding for this data is also not feasible as Smart Contract Accounts must
+avoid any ambiguity about the meaning of the received data.
+
+Using a different encoding will not benefit from the fact that ABI decoding logic is already built into
+Solidity programming language and ships with most of deployed smart contracts.
+
+### Allowing regular code execution flows to access `kind_abstraction` sections
+
+The code in a `kind_abstraction` does not have any difference
+There is absolutely no difference between `kind_abstraction (0x05)` and `kind_code (0x02)` sections outside the
+context of Native Account Abstraction transactions.
+
+The `kind_abstraction` specifies an alternative starting point for a system frame execution, but after that
+code from different sections can be accessed as well.
+
+There seems to be no reason to isolate `kind_abstraction` code section in any way.
+
+### Using `action` identifier instead of different kind markers
+
+As an alternative, we could define multiple kind markers, for example `kind_validate (0x05)`, `kind_validate_paymaster`,
+`kind_deploy_account` and `kind_paymaster_post_transaction`.
+
+However, this approach would fill the kind markers namespace with very particular types of code sections with little
+practical difference between them. This approach could also complicate creation of multi-role smart contracts,
+for example a contract sharing the `Paymaster` and `Deployer` functionality.
+
+Defining `action` field serves as a language-agnostic method selector, and in the future could be extended with
+additional Account Abstraction related functionality.
+
+## Backwards Compatibility
+
+An EOF contract with `kind_abstraction` section is not valid according to EIP-3540 and cannot exist on-chain
+before this proposal is implemented.
+
+The introduction of `kind_abstraction` will break an assumption that a contract code can only have a single
+execution starting point, which might confuse some developer tooling that relies on this assumption.
+
+## Security Considerations
+
+A contract with a `kind_abstraction` code section can be immediately identified as a Native Account Abstraction entity.
+This is a significant improvement over ERC-4337 and RIP-7560 where entities are not explicitly marked.
```

Comment:
and with different kinds for different entities, this becomes even safer.  An account that *also* has a `kind_paymaster_validation` will stand out as a likely rugpull attempt unless there's a good explanation for having that section.